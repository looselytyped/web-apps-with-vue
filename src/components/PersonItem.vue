<template>
  <div>
    <v-list-item :key="friend.id">
      <v-list-item-content>
        <v-list-item-title>
          {{ fullName }}
        </v-list-item-title>
      </v-list-item-content>

      <div class="flex-grow-1"></div>

      <v-btn icon>
        <v-icon class="edit">mdi-pencil</v-icon>
      </v-btn>
      <v-btn icon @click.stop="like()">
        <v-icon class="fav" :color="friend.fav ? 'red' : 'grey'">
          mdi-heart
        </v-icon>
      </v-btn>
    </v-list-item>
    <v-divider v-if="!last"></v-divider>
  </div>
</template>

<script>
export default {
  props: {
    friend: {
      type: Object,
      required: true
    },
    last: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  methods: {
    like() {
      // TODO: Do NOT modify props! Fix this
      this.$emit("friend-liked", this.friend);
    }
  },
  computed: {
    fullName() {
      return `${this.friend.firstName} ${this.friend.lastName}`;
    }
  }
};
</script>

<style></style>
