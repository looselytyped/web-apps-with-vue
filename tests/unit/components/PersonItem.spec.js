import { mount, shallowMount, createLocalVue } from "@vue/test-utils";
import Vuetify from "vuetify";
import PersonItem from "@/components/PersonItem.vue";

describe("PersonItem", () => {
  let localVue;
  let vuetify;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
  });

  it("should display a Person", () => {
    expect.assertions(1);
    const wrapper = shallowMount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.text()).toContain("Raju Gandhi");
  });

  it("should display a divider if its NOT the last friend", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: false
      }
    });

    expect(wrapper.find(".v-divider").exists()).toBe(true);
  });

  it("should NOT display a divider if its the last friend", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.find(".v-divider").exists()).toBe(false);
  });

  it("should highlight the favorite icon if the friend is a fav", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: false
      }
    });
    expect(wrapper.find("i.v-icon.red--text").exists()).toBe(true);
  });

  it("should grey out the favorite icon if the friend is NOT a fav", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });
    expect(wrapper.find("i.v-icon.grey--text").exists()).toBe(true);
  });

  it("should display the full name", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });
    const vm = wrapper.vm;
    expect(vm.fullName).toBe("Raju Gandhi");
  });

  it("should emit liked event when friend is liked", () => {
    expect.assertions(1);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });

    wrapper.vm.like = jest.fn();

    wrapper.find("i.v-icon.fav").trigger("click");
    expect(wrapper.vm.like).toHaveBeenCalledTimes(1);
  });

  it("should emit 'friend-liked' when 'like' handler is invoked", () => {
    expect.assertions(2);
    const wrapper = mount(PersonItem, {
      localVue,
      vuetify,
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });

    wrapper.vm.like();
    const emitted = wrapper.emitted();
    const e = emitted["friend-liked"];
    expect(e).toBeTruthy();
    expect(e.length).toBe(1);
  });

  it("should render correctly", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
