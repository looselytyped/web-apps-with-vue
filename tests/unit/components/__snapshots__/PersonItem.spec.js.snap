// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonItem should render correctly 1`] = `
<div>
  <div
    class="v-list-item theme--light"
    tabindex="-1"
  >
    <div
      class="v-list-item__content"
    >
      <div
        class="v-list-item__title"
      >
        
        Raju Gandhi
      
      </div>
    </div>
     
    <div
      class="flex-grow-1"
    />
     
    <button
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate edit mdi mdi-pencil theme--light"
        />
      </span>
    </button>
     
    <button
      class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate fav mdi mdi-heart theme--light red--text"
        />
      </span>
    </button>
  </div>
   
  <!---->
</div>
`;
